ext/ddtrace_profiling_native_extension/stack_recorder.c

```diff
@@ -148,3 +325,95 @@ static void *call_serialize_without_gvl(void *call_args) {
 void enforce_recorder_instance(VALUE object) {
   Check_TypedStruct(object, &stack_recorder_typed_data);
 }
+
+static struct active_slot_pair sampler_lock_active_profile(struct stack_recorder_state *state) {
+  int error;
+
+  for (int attempts = 0; attempts < 2; attempts++) {
```

Comment:
Do you think it makes sense to yield between retries? https://man7.org/linux/man-pages/man2/sched_yield.2.html (or similar)

As a related question: is 2 attempts enough to stop profiling for the lifetime of the application?

Reply:
> Do you think it makes sense to yield between retries? https://man7.org/linux/man-pages/man2/sched_yield.2.html (or similar)

No, because the retry is here only to protect the following situation (time flows down)

| sampler thread                | serializer thread                    |
|-------------------------------|--------------------------------------|
|                               | (starting state: slot one is locked) |
| trylock(slot one) --> busy    |                                      |
|                               | serialize gets called                |
|                               | unlock(slot one)                     |
|                               | lock(slot two)                       |
| trylock(slot two) --> busy    |                                      |

Because serialization happens once a minute, even if we get extremely unlucky and this happens, then retrying immediately after is enough -- the action from another thread already happened.

Furthermore, the sampler thread will be holding on to the global vm lock, so the last thing we want is for it to be blocked. This is why we even use trylock on the sampler thread path -- those either grab the lock or don't, there's no waiting.

> As a related question: is 2 attempts enough to stop profiling for the lifetime of the application?

That's a good question. I'm thinking thinking it's nice to have stronger assurances that things are as we expect. As I mentioned above, if two times in a row you failed to lock the locks, then something unexpected is happening, and I'm thinking it's better to be on the safe side and avoid impacting the application. It's easier to impact the application from the native side so I think we need to be extra careful not to do so.

(Note that we can always catch the exception and try again later, so technically the "stopping profiling" behavior won't even live in this component)

Reply:
All clear!